network/policy/policy_windows.go

```diff
@@ -45,22 +50,67 @@ type KVPairRoute struct {
 	NeedEncap         bool          `json:"NeedEncap"`
 }
 
+var ValidWinVerForDnsNat bool
+
+func init() {
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Microsoft\Windows NT\CurrentVersion`, registry.QUERY_VALUE)
+	if err == nil {
+		defer k.Close()
+
+		cb, _, err := k.GetStringValue("CurrentBuild")
+		if err == nil {
+			winVer, err := strconv.Atoi(cb)
+			if err == nil {
+				ValidWinVerForDnsNat = winVer >= win1903Version
+			}
+		}
+	}
+
+	if err != nil {
+		log.Errorf(err.Error())
+		panic(err)
+	}
+}
+
 // SerializePolicies serializes policies to json.
-func SerializePolicies(policyType CNIPolicyType, policies []Policy, epInfoData map[string]interface{}) []json.RawMessage {
-	var jsonPolicies []json.RawMessage
+func SerializePolicies(policyType CNIPolicyType, policies []Policy, epInfoData map[string]interface{}, enableSnatForDns, enableMultiTenancy bool) []json.RawMessage {
+	var (
+		jsonPolicies    []json.RawMessage
+		addDnsNatPolicy bool
+	)
 	for _, policy := range policies {
 		if policy.Type == policyType {
 			if isPolicyTypeOutBoundNAT := IsPolicyTypeOutBoundNAT(policy); isPolicyTypeOutBoundNAT {
-				if serializedOutboundNatPolicy, err := SerializeOutBoundNATPolicy(policy, epInfoData); err != nil {
-					log.Printf("Failed to serialize OutBoundNAT policy")
+				if enableMultiTenancy {
+					if enableSnatForDns {
+						addDnsNatPolicy = ValidWinVerForDnsNat
+						if serializedOutboundNatPolicy, err := SerializeOutBoundNATPolicy(policy, epInfoData); err != nil {
+							log.Printf("Failed to serialize OutBoundNAT policy")
+						} else {
+							jsonPolicies = append(jsonPolicies, serializedOutboundNatPolicy)
+						}
+					}
 				} else {
-					jsonPolicies = append(jsonPolicies, serializedOutboundNatPolicy)
+					if serializedOutboundNatPolicy, err := SerializeOutBoundNATPolicy(policy, epInfoData); err != nil {
+						log.Printf("Failed to serialize OutBoundNAT policy")
+					} else {
+						jsonPolicies = append(jsonPolicies, serializedOutboundNatPolicy)
+					}
```

Comment:
code repetition. I recommend to redesign the whole code..there are too many ifs and else..it can be modeled better or move to separate function

Reply:
updated